src/Qir/Runtime/lib/QIR/QubitManager.cpp

```diff
@@ -0,0 +1,465 @@
+        return;
+    }
+
+    if (addToFront)
+    {
+        sharedQubitStatusArray[id] = firstElement; // The new element will point to the former first element.
+        firstElement = id; // The new element is now the first in the chain.
+    } else
+    {
+        sharedQubitStatusArray[lastElement] = id; // The last element will point to the new element.
+        sharedQubitStatusArray[id] = NoneMarker; // The new element will end the chain.
+        lastElement = id; // The new element will be the last element in the chain.
+    }
+}
+
+// TODO: Set status to the taken qubit here. Then counting is reasonable here, but not possible?
+// TODO: Rename 'RemoveQubitFromFront'?
+int QubitListInSharedArray::TakeQubitFromFront(QubitIdType* sharedQubitStatusArray)
+{
+    // First element will be returned. It is 'NoneMarker' if the list is empty.
+    int result = firstElement;
+
+    // Advance list start to the next element if list is not empty.
+    if (!IsEmpty())
+    {
+        firstElement = sharedQubitStatusArray[firstElement]; // The second element will be the first.
+    }
+
+    // Drop pointer to the last element if list becomes empty.
+    if (IsEmpty())
+    {
+        lastElement = NoneMarker;
+    }
+
+    return result;
+}
+
+void QubitListInSharedArray::MoveAllQubitsFrom(QubitListInSharedArray& source, QubitIdType* sharedQubitStatusArray)
+{
+    // No need to do anthing if source is empty.
+    if (source.IsEmpty())
+    {
+        return;
+    }
+
+    if (this->IsEmpty())
+    {
+        // If this list is empty, we'll just set it to the source list.
+        lastElement = source.lastElement;
+    } else
+    {
+        // Attach source at the beginning of the list if both lists aren't empty.
+        sharedQubitStatusArray[source.lastElement] = firstElement; // The last element of the source chain will point to the first element of this chain.
+    }
+    firstElement = source.firstElement; // The first element from the source chain will be the first element of this chain.
+
+    // Remove all elements from source.
+    source.firstElement = NoneMarker;
+    source.lastElement = NoneMarker;
+}
+
+
+//
+// RestrictedReuseArea
+//
+
+RestrictedReuseArea::RestrictedReuseArea(QubitListInSharedArray freeQubits)
+{
+    //FreeQubitsReuseProhibited = QubitListInSharedArray(); // This is initialized by default.
+    FreeQubitsReuseAllowed = freeQubits; // Default shallow copying.
+}
+
+
+//
+// CRestrictedReuseAreaStack
+//
+
+void CRestrictedReuseAreaStack::PushToBack(RestrictedReuseArea area)
+{
+    FailIf(Count() >= std::numeric_limits<int>::max(), "Too many nested restricted reuse areas.");
+    this->insert(this->end(), area);
+}
+
+RestrictedReuseArea CRestrictedReuseAreaStack::PopFromBack()
+{
+    FailIf(this->empty(), "Cannot remove element from empty set.");
+    RestrictedReuseArea result = this->back();
+    this->pop_back();
+    return result;
+}
```

Comment:
(FYI)
From the exception-safety and thread-safety point of view it is important that `Pop()` does not return a value.
I don't remember the details well enough to explain, but, if you are interested, they can be found in
* [3.2.3 Spotting race conditions inherent in interfaces](https://www.bogotobogo.com/cplusplus/files/CplusplusConcurrencyInAction_PracticalMultithreading.pdf)
* [[exc_h:fsos]](https://github.com/kuzminrobin/code_review_notes/blob/master/article_list.md)

Reply:
This is not part of a public interface and is used on an object not intended for multithreading. So this is acceptable.
Yes, in general, this code would definitely be unacceptable for multithreading. Also, if part of the public interface, it would not be a good idea to have such a function as it would be impossible to tell if the element was actually removed in case of exception thrown.